### PR TITLE
Add route to get member cluster node

### PR DESCRIPTION
**Pull request by Heylosky**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Currently our company is using karmada as a company-level cluster management platform. We plan to display more information on the dashboard and add some new features, such as the ability to display member-cluster-node information. This PR adds the ability to obtain member-cluster-node information through the karmada-dashboard-api component, in preparation for adding a UI page.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
In our usage scenario, the member cluster is isolated from the main cluster. We hope to display more information about the member cluster through the dashboard, such as deployment, pod, namespace, etc., not just the resources deployed by karmada. Please consider whether it can be added to future plans.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Hi @Heylosky Thanks for spotting and doing this.
I wonder if you will attend this week's community meeting, we can talk about it there.
You can find the meeting link at https://github.com/karmada-io/dashboard?tab=readme-ov-file#meeting.

**Comment by Heylosky** [comment]
> Hi @Heylosky Thanks for spotting and doing this. I wonder if you will attend this week's community meeting, we can talk about it there. You can find the meeting link at https://github.com/karmada-io/dashboard?tab=readme-ov-file#meeting.

OK.

File: cmd/api/app/routes/node/handler.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/karmada-io/dashboard/cmd/api/app/router"
+	"github.com/karmada-io/dashboard/cmd/api/app/types/common"
+	"github.com/karmada-io/dashboard/pkg/client"
+	"github.com/karmada-io/dashboard/pkg/resource/node"
+)
+
+func handleGetClusterNode(c *gin.Context) {
+	karmadaClient := client.InClusterClientForKarmadactlApiServer(c.Param("clustername"))
```
Comment (warjiang) [inline cmd/api/app/routes/node/handler.go:12]: maybe we need to check the existence of member cluster，after init karmada client, you can check with following implementation:
```go
_, err = client.ClusterV1alpha1().Clusters().Get(context.TODO(), clusterName, metav1.GetOptions{})
if err != nil {
    // member cluster not exist
}
```

File: pkg/client/init.go [hunk 2]
```diff
@@ -3,22 +3,26 @@ package client
...
-	inClusterClientForKarmadaApiServer kubeclient.Interface
+	kubernetesRestConfig                  *rest.Config
+	kubernetesApiConfig                   *clientcmdapi.Config
+	inClusterClient                       kubeclient.Interface
+	karmadaRestConfig                     *rest.Config
+	karmadaApiConfig                      *clientcmdapi.Config
+	inClusterKarmadaClient                karmadaclientset.Interface
+	inClusterClientForKarmadaApiServer    kubeclient.Interface
+	inClusterClientForKarmadactlApiServer kubeclient.Interface
+	clusterHost                           string
```
Comment (warjiang) [inline pkg/client/init.go:25]: the reason for we need to add `clusterHost` is that in the implementation, we modify the `rest.Config` for karmada. I think we should not modify the `rest.Config` for karmada, the recommend way is deepcopy of `rest.Config`.

File: pkg/client/init.go [hunk 3]
```diff
@@ -233,6 +239,29 @@ func InClusterClientForKarmadaApiServer() kubeclient.Interface {
 	return inClusterClientForKarmadaApiServer
 }
 
+func InClusterClientForKarmadactlApiServer(clustername string) kubeclient.Interface {
```
Comment (warjiang) [inline pkg/client/init.go:242]: maybe `InClusterClientForMemberCluster` is good ~😃

Comment (warjiang) [reply pkg/client/init.go:242]: split client for karmada apiserver and member cluster

File: pkg/client/init.go [hunk 4]
```diff
@@ -233,6 +239,29 @@ func InClusterClientForKarmadaApiServer() kubeclient.Interface {
...
+	if !isKarmadaInitialized() {
+		return nil
+	}
+	karmadaCtlRestConfig, _, err := GetKarmadaConfig()
+	if strings.Contains(karmadaCtlRestConfig.Host, clustername) {
+		return inClusterClientForKarmadactlApiServer
+	} else {
+		karmadaCtlRestConfig.Host = clusterHost + "/apis/cluster.karmada.io/v1alpha1/clusters/" + clustername + "/proxy"
```
Comment (warjiang) [inline pkg/client/init.go:250]: maybe this way is better:
```go
const proxyURL = "/apis/cluster.karmada.io/v1alpha1/clusters/%s/proxy/"
memberAPIServer := karmadaRestConfig.Host + fmt.Sprintf(proxyURL, clusterName)
```

File: pkg/client/init.go [hunk 5]
```diff
@@ -233,6 +239,29 @@ func InClusterClientForKarmadaApiServer() kubeclient.Interface {
...
+	if err != nil {
+		klog.ErrorS(err, "Could not get karmadactl restConfig")
+		return nil
+	}
+	c, err := kubeclient.NewForConfig(karmadaCtlRestConfig)
+	if err != nil {
+		klog.ErrorS(err, "Could not init kubernetes in-cluster client for karmadactl apiserver")
+		return nil
+	}
+	inClusterClientForKarmadactlApiServer = c
```
Comment (warjiang) [inline pkg/client/init.go:261]: client for member cluster need multi instances, maybe use map to store the client ?

File: cmd/api/app/routes/node/handler.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	if err != nil {
+		common.Fail(c, err)
+		return
+	}
+	common.Success(c, result)
+}
+
+func init() {
+	r := router.V1()
+	r.GET("/node/:clustername", handleGetClusterNode)
```
Comment (warjiang) [inline cmd/api/app/routes/node/handler.go:24]: and we should add prefix for member cluster resource, currently resources name with / prefix represent the resource in the control plane, which means we get the resource from karmada apiserver. for member cluster's resources, need prefix like `/member/:clustername`, so the route should be corrected as `/member/:clustername/node`, any suggestions?

**Comment by warjiang** [comment]
/lgtm
/approve
thanks @Heylosky

File: pkg/client/init.go [hunk 7]
```diff
@@ -233,6 +254,43 @@ func InClusterClientForKarmadaApiServer() kubeclient.Interface {
 	return inClusterClientForKarmadaApiServer
 }
 
+func InClusterClientForMemberCluster(clusterName string) kubeclient.Interface {
```
Comment (warjiang) [inline pkg/client/init.go:257]: good 👍

**Pull request merged**